consensus/state_sync/state_sync.go

```diff
@@ -0,0 +1,104 @@
+package state_sync
+
+import (
+	typesCons "github.com/pokt-network/pocket/consensus/types"
+	cryptoPocket "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+type StateSyncModule interface {
+	modules.Module
+
+	// -- Constructor Setter Functions --
+
+	// `HandleBlock` function contract:
+	// - Create a Utility Context
+	// - Block.ValidateBasic()
+	// - Consensus Module Replica Path
+	//   - Prepare Block:  utilityContext.SetProposalBlock(block)
+	//   - Apply Block:    utilityContext.ApplyBlock(block)
+	//   - Validate Block: utilityContext.AppHash == Block.AppHash
+	//   - Store Block:    consensusModule.CommitBlock()
+	SetHandleBlock(HandleBlock func(block *typesCons.Block) error)
```

Comment:
The callback paradigm feels very `cosmosy` to me and I'd like to avoid it if possible.

The application specific bus currently supports two approaches of communication:
1. Synchronous (e.g. `m.GetBus().GetModule().CallFoo(...)`); currently used in most places
2. Asynchoronous (e.g. `node.GetBus().PublishEventToBus(e)`); currently used by P2P to send messages to consensus

#2 flows through the following code path:
```go
	// While loop lasting throughout the entire lifecycle of the node to handle asynchronous events
	for {
		event := node.GetBus().GetBusEvent()
		if err := node.handleEvent(event); err != nil {
			log.Println("Error handling event: ", err)
		}
	}
```

The idea here was that it's easy to log, capture errors, telemetry, etc for this that are async in a centralized manner.

I would prefer that this be changed to:

```
HandleBlock(*typesCons.Block) error
```

This way we have optionality on it being called through (1) OR (2) and not need to manage a bunch of async callbacks comes and going to different places.

Reply:
Hmmm `triggered` by **Cosmosy** xd

Okay so I see your point, however - technically speaking - callbacks are more package agnostic and could be used in another application without requiring an `AppBus`. 

However, to your point, we are building for Pocket. 

WYT?

Reply:
Assuming you're for the bus here; 

How would you expect to reflect using the bus `In_This_Commit` when we just have interfaces.

Describe it in a comment?

Reply:
tl;dr `StateSyncModule` would have a function called `HandleStateSyncMessage(typesCons.StateSyncMessage) error` that is expected to be called by `ConsensusModule.HandleMessage`, and `In_This_Commit`, we would have a protobuf called

```proto
enum StateSyncMessageType {
...
    STATE_SYNC_PROVIDE_BLOCK = N;
...
}

message StateStateProvideBlockMessage {
    Block block = 1;
}

message StateSyncMessage {
    StateSyncMessageType type = 1;
    google.protobuf.Any message = 2;
    ...
}
```

That gives us the ability to send/receive various types of state sync messages between nodes.

If we go with this approach, it might require changing the flow a bit, because there's room for making the type "Provide, Request, etc...", but also makes it easier to adapt in the future within a switch case statement we implement.

--- 

The consensus module has a function called `HandleMessage(*anypb.Any) error`.

That function is implemented like so:

```go
func (m *consensusModule) HandleMessage(message *anypb.Any) error {
	m.m.Lock()
	defer m.m.Unlock()
	switch message.MessageName() {
	case HotstuffMessage:
		msg, err := codec.GetCodec().FromAny(message)
		if err != nil {
			return err
		}
		hotstuffMessage, ok := msg.(*typesCons.HotstuffMessage)
		if !ok {
			return fmt.Errorf("failed to cast message to HotstuffMessage")
		}
		if err := m.handleHotstuffMessage(hotstuffMessage); err != nil {
			return err
		}
	case UtilityMessage:
		panic("[WARN] UtilityMessage handling is not implemented by consensus yet...")
	default:
		return typesCons.ErrUnknownConsensusMessageType(message.MessageName())
	}

	return nil
}
```

Which uses a protobuf like so:

```proto
message HotstuffMessage  {
    HotstuffMessageType type = 1;
    uint64 height = 2;
    ...
```

My assumption is that when we call `RunSyncMode`, it would do things like `m.GetBus().GetP2P().Send(<GiveMeBlockX>)` and eventually `StateSyncModule.HandleMessage(msg)` would get triggered asynchronously.

This is quite different from what we have right now, but I think might give the implementation more flexibility when we get there and more extendability in the future.

Reply:
Disagree with the needs for structures or protobuf `IN_THIS_COMMIT` - but I understand the point to be:

TL;DR - 

```golang
// Replace
SetHandleBlock(HandleBlock func(block *typesCons.Block) error)

// With
HandleStateSyncMessage(msg BlockResponseMessage)
```

Also - any thoughts on my feedback:
> callbacks are more package agnostic and could be used in another application without requiring an AppBus.

Reply:
> callbacks are more package agnostic and could be used in another application without requiring an AppBus.

I can see myself arguing both sides, but ultimately it's about tradeoffs and the overall design of the system.
Every system needs have the ability to send message synchronously and asynchronously.

**Synchronous**:
   * Passing data around manually
   * Having a Bus and calling functions similar to what we have
   * A dependency injection framework
   * A design pattern similar to react's reflex

**Asynchronous**:
   * Pub/sub
   * Callbacks
   * Promises
   * Async/await

Re callbacks:
- Pro: it's asynchronous, and there is no "central nervous system" (i.e. bus)
- Con: are all the callbacks going to be set and/or defined in one place or will it be dispersed?

Re dependency injection:
- Pro: a single place defines everything that we inject into the system
- Con: Customizing injections on a per module/submodule place can get difficult, and it's hard for different modules to be independent.

Re pub/sub:
- Pro: A really great design (in theory) that's completely independent from the core logic and everything is defined via topics and messages.
- Con: These things are really hard to debug, and come with tradeoffs (at least once, at most once, exactly once, etc...) and are often hard to debug


This is just a brought set of points / ideas, but I can go into more depth if you'd like.

My biggest concern, and the reason I feel strongly against callbacks, is because having too many different patterns can lead to requiring different types of tooling for debugging/visibility, different workflows, different types of test suites, different mindsets/approaches to different parts of the codebase, and overall disparity when trying to understand how the whole system works

The app bus in this case, is similar to some sort of object that sets the callback. Calling the function based on an event is a continuous runtime thing, whereas setting the callback is a one time event when the application starts up. My preference is the former is because I've found it to be much easier to debug. There is a single place where you can add logging, a breakpoint, etc, and take it from there.

---

I wanted to keep this "short" but happy to provide a more formulated argument around this. Lmk

Reply:
> Disagree with the needs for structures or protobuf IN_THIS_COMMIT

That's fair. I don't think we have a great definition of what should & shouldn't go into the interface design PR (e.g. should we include protobufs or not), so it's up to the architect.

We could add a comment saying "Implementation may need to define a set of protobufs" and I think identifying them during implementation will also be easier and help reduce the scope of this PR independently.